lib/go-tc/v13/federations.go

```diff
@@ -0,0 +1,62 @@
+package v13
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Other endpoints define their own alert. Why not just use tc.Alert?
+type FederationAlert struct {
+	Level string `json:"level"`
+	Text  string `json:"text"`
+}
+
+type CDNFederationResponse struct {
+	Response []CDNFederation `json:"response"`
+}
+
+type CreateCDNFederationResponse struct {
+	Response CDNFederation     `json:"response"`
+	Alerts   []FederationAlert `json:"alerts"`
+}
+
+type UpdateCDNFederationResponse struct {
+	Reponse CDNFederation     `json:"response"`
+	Alerts  []FederationAlert `json:"alerts"`
+}
+
+type DeleteCDNFederationResponse struct {
+	Alerts []FederationAlert `json:"alerts"`
+}
+
+type CDNFederation struct {
+	ID          *int    `json:"id" db:"id"`
+	CName       *string `json:"cname" db:"cname"`
+	Ttl         *int    `json:"ttl" db:"ttl"`
+	Description *string `json:"description" db:"description"`
+
+	//omitempty only works with primitive types and pointers
+	*DeliveryServiceIDs `json:"deliveryService,omitempty"`
+
+	//Extra datapoint for ease in Read function (cannot Scan after StructScan)
```

Comment:
I'd prefer this to be named `CDNName` rather than just `Name`.

Also can you elaborate on "cannot Scan after StructScan"?

Reply:
A struct scan allows you to scan in such a way that you don't need to scan in order, which works by specifying db tags. After using struct scan, I guess the order cannot be guaranteed, so Scan (which works by order) will not work.

Also, I think a struct scan assumes that everything in the scan queue needs to fit into the struct.

When I had the Nullable and non-nullable structs I think this made a little more sense. Now, I think it'd be best to simply do a sequential scan of all the fields.

Reply:
```
//Option 1: (get data, and also fetch cdn information)
SELECT federation.id as id, cname, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice
RIGHT JOIN cdn ON cdn.id = cdn_id

//Option 2: (get data, then check if cdn exists)
SELECT federation.id as id, cname, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice

SELECT id from cdn WHERE cdn.name = :name
```

As far as splitting queries is concerned, would there be a big difference in efficiency? I initially did option 1 because additional database queries could be more expensive, but now I think option 2 would be best, since it would also allow me to scan into the struct directly for the first query.

Reply:
Yeah, I don't think one extra DB call per GET is really a big deal. Premature optimization is the root of all evil :). Prefer readability/simplicity to start.

Reply:
Why does it need a RIGHT JOIN? The `deliveryservice` table `cdn_id` is NOT NULL, can't you:
```
SELECT federation.id as id, cdn.name as cdn_name, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice
JOIN cdn ON cdn.id = deliveryservice.cdn_id
```
?

I would vote to avoid the StructScan and use a standard `sql.Tx.Scan`, before changing the query. IMO the single query is fast and readable, we shouldn't use suboptimal SQL to accommodate the limitations of the sqlx library. A `Scan(&f.ID, &f.CName, &f.TTL, &s.Description, &s.DeliveryServiceID, &s.XMLID, &cdnID)` isn't that hard to read.

I'm also -1 on putting `json:"-"` tags in `tc` structs, just to accommodate sqlx deficiencies. Our structs are our public interface, for anyone using Go or the Go TO client, as well as all our products that use Go and TO data. We should prioritize keeping that interface clean.

Reply:
Maybe part of the problem is that we're trying to handle two GET endpoints (`cdns/:name/federations` and `cdns/:name/federations/:fedId`) in the same `Read` handler. Perhaps it would make more sense to split this into two handlers to make them more clear, as the latter doesn't even really care about `cdn_id`.

Reply:
It doesn't need a right join if you separate out the queries. What I was trying to do was return federation data and cdn data together. That way, I could check to see if the requested cdn existed. Specifying a regular join gets the correct result, but if I got 0 rows I can't tell if thats because the cdn doesn't exist.

Reply:
BTW, in option 2 the second query would only be run if the first query has 0 rows.

Reply:
@rawlinp The problem here has to do more with fetching the cdn name. The id can be included or excluded elegantly. All I need to do is conditionally assign a different query (one that I haven't shown), and it should work.

Reply:
I could just not return a "Resource not found" error inside my handler, but then the behavior would change.

Reply:
>It doesn't need a right join if you separate out the queries.

>if I got 0 rows I can't tell if thats because the cdn doesn't exist.

It doesn't need a RIGHT JOIN as a single query. The cdn must exist: there's a foreign key constraint, and a not-null constraint.

+1 on using different functions. The ID handler can use the CRUDer abstractions, and the name handler can use a native http.HandlerFunc. I believe we have several endpoints already doing that.


Using field mapping like sqlx, and frameworks like ORMs, are always less powerful and expressive than SQL. SQL is expressive, powerful, and readable, as a 4th generation programming language. Likewise, objects aren't relational, and shouldn't be. We shouldn't try to shove a square peg into a round hole, when the data we need doesn't cleanly map to objects. We should use fast, clean, readable SQL; and map that to our objects afterwards. 

The same goes for HTTP/JSON. Clean APIs, clean objects, and clean SQL, with a bit of mapping code where necessary, is better than trying to hack any of those to fit the other. JSON, Objects, and SQL don't map perfectly; but they're all great in their respective contexts. Using each properly in its own context is much better—easier to learn, use, read, write, and generally work with—than trying to hack them to fit each-other, just to avoid a bit of mapping code between them. The mapping code isn't super-pretty, but it's far better than making the API/Object/SQL difficult to use and read everywhere it's used.

But that's just my opinion.

Reply:
Ah, sorry, I didn't understand that your issue was differentiating whether the CDN exists, to return different errors for nonexistent CDNs and nonexistent federations.

For that case, yeah, multiple queries is fine. It's not ideal, but that will make the SQL simpler. I thought you needed the ID. In fact, we already have functions for that, moving one to `dbhelpers/db_helpers.go` wouldn't be a bad idea.

https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/crconfig/servers.go#L366 

Though, you can still get that in a single IMO not-inelegant query

```
select distinct(cdn.name), f.cname, f.description
from cdn
left outer join deliveryservice as d on d.cdn_id = cdn.id
left outer join federation_deliveryservice as fd on fd.deliveryservice = d.id
left outer join federation as f on f.id = fd.federation
where cdn.name = $1;
```

Which will yield 1 row with the CDN and null federation info, if the CDN exists with no federations, and no rows if the CDN doesn't exist.

But, I wouldn't object to either way.